server/storage/backend/backend.go

```diff
@@ -477,29 +477,38 @@ func (b *backend) defrag() error {
 	b.readTx.Lock()
 	defer b.readTx.Unlock()
 
-	b.batchTx.unsafeCommit(true)
-
-	b.batchTx.tx = nil
-
 	// Create a temporary file to ensure we start with a clean slate.
 	// Snapshotter.cleanupSnapdir cleans up any of these that are found during startup.
 	dir := filepath.Dir(b.db.Path())
 	temp, err := os.CreateTemp(dir, "db.tmp.*")
 	if err != nil {
 		return err
 	}
+	defer temp.Close()
```

Comment:
Note that the temp file will be renamed after the defragmentation operation. I think we should close the temp file right before renaming. On Linux, it isn't a problem to rename a file when it's still open; but it's a problem on some other platform, i.e. Windows (although it's tier 3 for now).

https://github.com/etcd-io/etcd/blob/bd8896364a1d9adb554b538232004e5b2d6c850a/server/storage/backend/backend.go#L537

I agree that this `defer temp.Close()` still makes sense, but we need to guarantee it's only called once using `sync.Once`, although it isn't a big deal to call it multiple times.

So suggest to revert this change and resolve it in a separate PR.

Reply:
Oh good catch, I think that slipped in from my testing, I didn't want to commit this honestly.

Reply:
My previous comment isn't correct, bbolt will close the file when closing bbolt instance. The file should have already been closed before renaming the file.

https://github.com/etcd-io/bbolt/blob/c2e4c77280fc2f47870995aadd6e15c6422c5af0/db.go#L712